rmp/src/decode/mod.rs

```diff
@@ -244,6 +244,95 @@ pub fn read_marker<R: RmpRead>(rd: &mut R) -> Result<Marker, MarkerReadError<R::
     Ok(Marker::from_u8(rd.read_u8()?))
 }
 
+/// Attempts to read an Extension struct from the given reader and to decode it as a Timestamp value.
+///
+/// According to the MessagePack specification, there are 3 types of Timestamp, 32, 64, and 96.
+///
+/// # Errors
+///
+/// This function will return `ValueReadError` on any I/O error while reading the Timestamp,
+/// except the EINTR, which is handled internally.
+///
+/// It also returns `ValueReadError::TypeMismatch` if the actual type is not equal with the
+/// expected one, indicating you with the actual type.
+///
+/// # Note
+///
+/// This function will silently retry on every EINTR received from the underlying `Read` until
+/// successful read.
+///
+/// # Examples
+///
+/// ```
+/// use rmp::Timestamp;
+///
+/// // FixExt4 with a type of -1 (0xff)
+/// let mut buf1 = [0xd6, 0xff, 0x66, 0xc1, 0xde, 0x7c];
+/// // FixExt8 with a type of -1 (0xff)
+/// let mut buf2 = [0xd7, 0xff, 0xee, 0x6b, 0x27, 0xfc, 0x66, 0xc1, 0xde, 0x7c];
+/// // Ext8 with a size of 12 (0x0c) and a type of -1 (0xff)
+/// let mut buf3 = [0xc7, 0x0c, 0xff, 0x3b, 0x9a, 0xc9, 0xff, 0x00, 0x00, 0x00, 0x00, 0x66, 0xc1, 0xde, 0x7c];
+///
+/// let ts1_expected = Timestamp::T32 { secs: 0x66c1de7c };
+/// let ts2_expected = Timestamp::T64 { nsecs: 0x3b9ac9ff, secs: 0x66c1de7c };
+/// let ts3_expected = Timestamp::T96 { nsecs: 0x3b9ac9ff, secs: 0x66c1de7c };
+///
+/// let ts1_result = rmp::decode::read_timestamp(&mut buf1.as_slice()).unwrap();
+/// let ts2_result = rmp::decode::read_timestamp(&mut buf2.as_slice()).unwrap();
+/// let ts3_result = rmp::decode::read_timestamp(&mut buf3.as_slice()).unwrap();
+///
+/// assert_eq!(ts1_expected, ts1_result);
+/// assert_eq!(ts2_expected, ts2_result);
+/// assert_eq!(ts3_expected, ts3_result);
+/// ```
+pub fn read_timestamp<R: RmpRead>(rd: &mut R) -> Result<Timestamp, ValueReadError<R::Error>> {
+    let marker = read_marker(rd)?;
+    let prefix = rd.read_data_i8()?;
+    match (marker, prefix) {
+        // timestamp 32
+        (Marker::FixExt4, -1) => {
+            let secs = rd.read_data_u32()? as i64;
+            Ok(Timestamp::T32 {
+                secs,
+            })
+        },
+        // timestamp 64
+        (Marker::FixExt8, -1) => {
+            let data = rd.read_data_u64()?;
+            // 30 bits fits in u32
+            let nsecs = (data >> 34) as u32;
+            if nsecs > 999_999_999 {
```

Comment:
(After checking the chrono crate... leap seconds are represented by a second with a max of 1_999_999_999 nanos, so this check is invalid.......... I hate time.)

I'm leaning towards gating this behind a feature and using chrono crate to avoid weird overlooks like this.

Reply:
> In timestamp 64 and timestamp 96 formats, nanoseconds must not be larger than 999999999

https://github.com/msgpack/msgpack/blob/master/spec.md#timestamp-extension-type

................. I guess I was right in terms of........... the msgpack protocol........ at least...... (time is hard)